cloud-controller-manager/do/certificates_test.go

```diff
@@ -0,0 +1,318 @@
+/*
+Copyright 2017 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/digitalocean/godo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+type kvCertService struct {
+	store         map[string]*godo.Certificate
+	reflectorMode bool
+}
+
+func (f *kvCertService) Get(ctx context.Context, certID string) (*godo.Certificate, *godo.Response, error) {
+	if f.reflectorMode {
+		return &godo.Certificate{
+			ID:   certID,
+			Type: certTypeLetsEncrypt,
+		}, newFakeOKResponse(), nil
+	}
+	lb, ok := f.store[certID]
+	if ok {
+		return lb, newFakeOKResponse(), nil
+	}
+	return nil, newFakeNotOKResponse(), newFakeNotFoundErrorResponse()
+}
+
+func (f *kvCertService) List(ctx context.Context, listOpts *godo.ListOptions) ([]godo.Certificate, *godo.Response, error) {
+	panic("not implemented")
+}
+
+func (f *kvCertService) Create(ctx context.Context, crtr *godo.CertificateRequest) (*godo.Certificate, *godo.Response, error) {
+	panic("not implemented")
+}
+
+func (f *kvCertService) Delete(ctx context.Context, certID string) (*godo.Response, error) {
+	panic("not implemented")
+}
+
+func newKVCertService(store map[string]*godo.Certificate, reflectorMode bool) kvCertService {
+	return kvCertService{
+		store:         store,
+		reflectorMode: reflectorMode,
+	}
+}
+
+func createServiceAndCert(lbID, certID, certType string) (*v1.Service, *godo.Certificate) {
+	c := &godo.Certificate{
+		ID:   certID,
+		Type: certType,
+	}
+	s := createServiceWithCert(lbID, certID)
+	return s, c
+}
+
+func createServiceWithCert(lbID, certID string) *v1.Service {
+	s := createService(lbID)
+	s.Annotations[annDOCertificateID] = certID
+	return s
+}
+
+func createService(lbID string) *v1.Service {
+	return &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test",
+			UID:  "foobar123",
+			Annotations: map[string]string{
+				annDOProtocol:        "http",
+				annoDOLoadBalancerID: lbID,
+			},
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Name:     "test",
+					Protocol: "TCP",
+					Port:     int32(443),
+					NodePort: int32(30000),
+				},
+			},
+		},
+	}
+}
+
+func Test_LBaaSCertificateScenarios(t *testing.T) {
+	testcases := []struct {
```

Comment:
refactor: `s/testcases/tests/` to be consistent with the naming style we employ on more recent tests.